zkvm/src/constraints.rs

```diff
@@ -89,21 +92,58 @@ impl Constraint {
         .map_err(|e| VMError::R1CSError(e))
     }
 
-    fn flatten<CS: r1cs::RandomizedConstraintSystem>(self, cs: &mut CS) -> r1cs::LinearCombination {
+    fn flatten<CS: r1cs::RandomizedConstraintSystem>(
+        self,
+        cs: &mut CS,
+    ) -> Result<(r1cs::LinearCombination, Option<Scalar>), r1cs::R1CSError> {
         match self {
-            Constraint::Eq(expr1, expr2) => expr1.to_r1cs_lc() - expr2.to_r1cs_lc(),
+            Constraint::Eq(expr1, expr2) => {
+                let assignment = expr1
+                    .eval()
+                    .and_then(|x| expr2.eval().map(|y| x.to_scalar() - y.to_scalar()));
```

Comment:
```suggestion
                    .and_then(|x| expr2.eval().map(|y| (x - y).to_scalar()));
```

this probably would be a bit easier to read.